com.unity.shadergraph/Editor/Data/Graphs/MaterialSlot.cs

```diff
@@ -259,8 +259,29 @@ public bool IsCompatibleWith(MaterialSlot otherSlot)
 
         public bool IsCompatibleStageWith(MaterialSlot otherSlot)
         {
+            var downStreamShaderSatge = ShaderStageCapability.All;
+            var connectedStage = new List<ShaderStageCapability>();
+            var fromNode = otherSlot.owner;
+            var downstreamNodeList = new List<AbstractMaterialNode>();
+            NodeUtils.GetDownsteamNodesForNode(downstreamNodeList, fromNode);
+            foreach (var n in downstreamNodeList)
+            {
+                if (n.GetType() == typeof(BlockNode))
+                {
+                    var inputSlots = n.GetInputSlots<MaterialSlot>();
+                    connectedStage.Add(inputSlots.First().stageCapability);
+                }
+                else
+                {
+                    downStreamShaderSatge = otherSlot.stageCapability;
+                }
+            }
+            if (!connectedStage.Contains(ShaderStageCapability.Fragment))
```

Comment:
I'm not sure if this is true so it's a question, is it possible to get around having to store the shader stages for all downstream nodes and directly check within the loop itself as you go? 

What I'm thinking is that instead you could directly check in the loop if the BlockNode is of type Vertex or of type Fragment if its encountered, and if so then you could directly set the downStreamShaderStage and then break out of the loop.

I see that connectedStage is only ever added to when you encounter a BlockNode, and it seems to me that you could only ever have BlockNodes of either Fragment or Vertex stageCapability in the list of downstream nodes correct?

Because we already do prevent you from connecting to a Vertex stage block node from a Fragment stage node (and vice versa), so it seems like you could check if either of those are encountered and set the downStreamShaderStage directly to that value when first encountered.

Let me know if that makes sense, might be completely wrong

Reply:
I find NodeUtils.GetEffectiveShaderStageCapability() seems are checking the effective shader stage downstream and upstream(if its a output slot). I tried and it seemed working! So I used it to check rather than what I wrote previously.